File: csharp/dotnetcore_quiz/Controllers/QuizController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using SecureFromScratch.Quiz.Model;
using SecureFromScratch.Quiz.Services;
using SecureFromScratch.Quiz.Safety;
using SecureFromScratch.Quiz.Config;

namespace SecureFromScratch.Quiz.Controllers
{
    [ApiController]
    [Route("/")]
    public class QuizController : ControllerBase
    {
        private readonly ILogger<QuizController> _logger;
        private readonly ILogger _answersLogger;
        private readonly MultipleChoiceQuestion _quizService;

        public QuizController(
            ILogger<QuizController> logger,
            ILoggerFactory loggerFactory,
            MultipleChoiceQuestion quizService)
        {
            _logger = logger;
            _answersLogger = loggerFactory.CreateLogger("Answers");
            _quizService = quizService;
        }




        private static string GenerateAnswerOption(string username, string optionText, bool isCorrect)
        {
            return $"<a href='/answer?username={username}&isCorrect={isCorrect}'>{optionText}</a><br>\n";
        }

        public record QuestionDetails(string Question, string[] Answers);

        [HttpGet("question")]
        public ContentResult GetQuestionDetails([FromQuery(Name = "username")] string rawUsername)
        {
            _logger.LogInformation("/question with name={Username}", rawUsername);

            var question = _quizService.GetQuestion();
            _logger.LogInformation("/question returning question: {Question}", question);

            var username = new FullName(rawUsername);
            var shuffledAnswers = _quizService.GenerateShuffledAnswers();
            var options = shuffledAnswers.Options;
            var correctAnswerIdx = shuffledAnswers.CorrectAnswerIdx;

            var html = "<html>" +
                       "<head><title>Answer The Daily Question</title></head>" +
                       "<body>" +
                       $"<h2>{question}</h2><br>\n" +
                       "Here are the possible answers, choose the appropriate link:<br>\n" +
                       GenerateAnswerOption(username.Username, options[0], correctAnswerIdx == 0) +
                       GenerateAnswerOption(username.Username, options[1], correctAnswerIdx == 1) +
                       GenerateAnswerOption(username.Username, options[2], correctAnswerIdx == 2) +
                       GenerateAnswerOption(username.Username, options[3], correctAnswerIdx == 3) +
                       "</body></html>";

            return Content(html, "text/html");
        }


        private string GenerateAnswerOption(object value, string v1, bool v2)
        {
            throw new NotImplementedException();
        }

        [HttpGet("answer")]
        public ContentResult SubmitAnswer([FromQuery(Name = "username")] string rawUsername, [FromQuery] bool isCorrect)
        {
            _logger.LogInformation("/answer with name={Username} and answer={Answer}", rawUsername, isCorrect);

            var username = new FullName(rawUsername);

            if (isCorrect)
                _answersLogger.LogWarning("Right: {Username}", username.Username);
            else
                _answersLogger.LogInformation("Wrong: {Username}", username.Username);

            var html = "<html>" +
                       "<head><title>Done!</title></head>" +
                       "<body>" +
                       $"<h2>Thank you for submitting, {username.Username}</h2>" +
                       "<a href='/index.html'>Start Again</a>" +
                       "</body></html>";

            return Content(html, "text/html");
        }

    }
}

File: java/web_quiz_v2/src/main/java/com/securefromscratch/quiz/controllers/QuizController.java
package com.securefromscratch.quiz.controllers;

import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.security.SecureRandom;
import java.util.logging.Logger;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import static j2html.TagCreator.*;

import com.securefromscratch.quiz.safety.FullName;
import com.securefromscratch.quiz.config.LoggerConfig;
import com.securefromscratch.quiz.model.QuestionAndAnswers.ShuffledAnswers;
import com.securefromscratch.quiz.services.MultipleChoiceQuestion;
import com.securefromscratch.quiz.utils.IsCorrectObfuscator;
import com.securefromscratch.quiz.services.AnswerResultsStore;

@RestController
@RequestMapping("/")
public class QuizController {
    private static final Logger LOGGER = LoggerConfig.genLogger(QuizController.class.getSimpleName());
    private static final Logger ANSWERS_LOGGER = LoggerConfig.genLogger("Answers");

    private final MultipleChoiceQuestion m_quizService;

    private static String generateAnswerOption(String a_username, String a_optionText, boolean a_isCorrect) {
        return String.format(
            "<a href='/answer?username=%s&isCorrect=%s'>" 
            + a_optionText 
            + "</a><br>\n", a_username, IsCorrectObfuscator.encode(a_isCorrect));
    }

    public record QuestionDetails(String question, String[] answers) { }
    @GetMapping("/question")
    public String getQuestionDetails(@RequestParam("username") FullName username) {
        LOGGER.info(String.format("/question with name=%s", username));
        LOGGER.info(String.format("/question returning question: %s", m_quizService.getQuestion()));
        ShuffledAnswers shuffledAnswers = m_quizService.generateShuffledAnswers();
        String[] options = shuffledAnswers.options() ;
        int correctAnswerIdx = shuffledAnswers.correctAnswerIdx();
        return "<html>"
            + "<head><title>Answer The Daily Question</title></head>"
            + "<body>"
            + "<h2>" + m_quizService.getQuestion() + "</h2><br>\n"
            + "Here are the possible answers, choose the appropriate link:<br>\n"
            + generateAnswerOption(username.get(), options[0], 0 == correctAnswerIdx)
            + generateAnswerOption(username.get(), options[1], 1 == correctAnswerIdx)
            + generateAnswerOption(username.get(), options[2], 2 == correctAnswerIdx)
            + generateAnswerOption(username.get(), options[3], 3 == correctAnswerIdx)
            + "</body></html>";
    }

    @GetMapping("/answer")
    public String submitAnswer(@RequestParam("username") FullName username, @RequestParam("isCorrect") String isCorrectAsBase64) throws IOException {
        LOGGER.info(String.format("/answer with name=%s and answer=%s", username, isCorrectAsBase64));

        boolean isCorrect = IsCorrectObfuscator.decode(isCorrectAsBase64);
        writeAnswerResultToFile(username, isCorrect);

        return html(
            head(title("Done!")),
            body(
                h2("Thank you for submitting, " + username),
                a("Start Again").withHref("/index.html")
            )
        ).render();
    }

    private final AnswerResultsStore m_resultsStore;

    public QuizController(MultipleChoiceQuestion quizService, AnswerResultsStore resultsStore) {
        this.m_quizService = quizService;
        this.m_resultsStore = resultsStore;
    }

    private void writeAnswerResultToFile(FullName username, boolean isCorrect) throws IOException {
        m_resultsStore.addLine(...); // DO NOT USE COLON OR NEWLINE!

        /*try (BufferedWriter writer = Files.newBufferedWriter(
                Paths.get("answers.txt"), 
                StandardOpenOption.APPEND, 
                StandardOpenOption.CREATE)) {
            if (isCorrect) {
                writer.write("Right: ");
            }
            else {
                writer.write("Wrong: ");
            }
            writer.write(username.get());
            writer.newLine();
        }*/
    }
}
